src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
I disagree - I'd need the same special casing (except in clone), except the switch would be on `if (predicate.parameterName === "this" && predicate.parameterIndex === 1)` which is wholly uglier. This-based type guards and identifier-based types guards get parsed differently, checked differently, and find what to narrow differently - they actually share very little implementation, and the safety afforded by splitting it into two types (and then using a union of the two as the signature member, rather than the base type) is very nice. I do need to fix `cloneTypePredicate` to use the type guard I made though - I hadn't yet made it when I was writing it and neglected to go back and update it.

Additionally, over the past few hours I've realized that `this`-based type guards need to be extended to be allowed as property values/getter return types (this was in the original proposal and seems very reasonable), further separating them from identifier-based predicates.

The main thing they share is the slot on `Signature` which stores them, as they're mutually exclusive, but the way they're handled at every step is very different.